File: script-dashboard/src/crypto/secrets.ts
import { pbkdf2 } from '@noble/hashes/pbkdf2.js';
import { sha256 } from '@noble/hashes/sha2.js';
import { gcm } from '@noble/ciphers/aes.js';

const SECRETS_STORAGE_KEY = 'dashaws-secrets-blob';
const PBKDF2_ITERATIONS = 200_000;
const KEY_LENGTH_BITS = 256;
const SALT_BYTES = 32;

const useSubtle = (() => {
  try {
    return !!(globalThis.crypto && globalThis.crypto.subtle);
  } catch {
    return false;
  }
})();

function encode(buf: Uint8Array): string {
  return btoa(String.fromCharCode(...buf));
}

function decode(str: string): Uint8Array<ArrayBuffer> {
  const bin = atob(str);
  const bytes = new Uint8Array(bin.length) as Uint8Array<ArrayBuffer>;
  for (let i = 0; i < bin.length; i++) {
    bytes[i] = bin.charCodeAt(i);
  }
  return bytes;
}

function randomBytes(length: number): Uint8Array<ArrayBuffer> {
  return globalThis.crypto.getRandomValues(new Uint8Array(length)) as Uint8Array<ArrayBuffer>;
}

// ── Web Crypto API (crypto.subtle) path ────────────────────────────────

async function deriveKeySubtle(password: string, salt: Uint8Array<ArrayBuffer>): Promise<CryptoKey> {
  const baseKey = await crypto.subtle.importKey(
    'raw',
    new TextEncoder().encode(password),
    'PBKDF2',
    false,
    ['deriveKey']
  );
  return crypto.subtle.deriveKey(
    {
      name: 'PBKDF2',
      salt,
      iterations: PBKDF2_ITERATIONS,
      hash: 'SHA-256',
    },
    baseKey,
    { name: 'AES-GCM', length: KEY_LENGTH_BITS },
    false,
    ['encrypt', 'decrypt']
  );
}

// ── Pure-JS fallback (@noble/hashes + @noble/ciphers) ─────────────────

function deriveKeyJS(password: string, salt: Uint8Array): Uint8Array {
  return pbkdf2(sha256, password, salt, { c: PBKDF2_ITERATIONS, dkLen: KEY_LENGTH_BITS / 8 });
}

function encryptSecretsJS(secrets: Record<string, string>, password: string): EncryptedBlob {
  const salt = randomBytes(SALT_BYTES);
  const key = deriveKeyJS(password, salt);
  const nonce = randomBytes(12);
  const plaintext = new TextEncoder().encode(JSON.stringify(secrets));
  const cipher = gcm(key, nonce);
  const encrypted = cipher.encrypt(plaintext);
  const hash = hashPasswordJS(password);
  return { iv: encode(nonce), data: encode(encrypted), salt: encode(salt), hash };
}

function decryptSecretsJS(blob: EncryptedBlob, password: string): Record<string, string> {
  const salt = decode(blob.salt);
  const key = deriveKeyJS(password, salt);
  const nonce = decode(blob.iv);
  const ciphertext = decode(blob.data);
  const cipher = gcm(key, nonce);
  const decrypted = cipher.decrypt(ciphertext);
  return JSON.parse(new TextDecoder().decode(decrypted));
}

function hashPasswordJS(password: string): string {
  const hash = sha256(new TextEncoder().encode(password));
  return encode(hash);
}

// ── Public API ─────────────────────────────────────────────────────────

export interface EncryptedBlob {
  iv: string;
  data: string;
  salt: string;
  hash: string;
}

export async function encryptSecrets(
  secrets: Record<string, string>,
  password: string
): Promise<EncryptedBlob> {
  if (useSubtle) {
    const salt = randomBytes(SALT_BYTES);
    const key = await deriveKeySubtle(password, salt);
    const iv = randomBytes(12);
    const plaintext = JSON.stringify(secrets);
    const encrypted = await crypto.subtle.encrypt(
      { name: 'AES-GCM', iv },
      key,
      new TextEncoder().encode(plaintext)
    );
    const hash = await hashPassword(password);
    return {
      iv: encode(iv),
      data: encode(new Uint8Array(encrypted)),
      salt: encode(salt),
      hash,
    };
  }
  return encryptSecretsJS(secrets, password);
}

export async function decryptSecrets(
  blob: EncryptedBlob,
  password: string
): Promise<Record<string, string>> {
  if (useSubtle) {
    const salt = decode(blob.salt);
    const key = await deriveKeySubtle(password, salt);
    const iv = decode(blob.iv);
    const data = decode(blob.data);
    const decrypted = await crypto.subtle.decrypt(
      { name: 'AES-GCM', iv },
      key,
      data
    );
    return JSON.parse(new TextDecoder().decode(decrypted));
  }
  return decryptSecretsJS(blob, password);
}

export async function hashPassword(password: string): Promise<string> {
  if (useSubtle) {
    const hash = await crypto.subtle.digest('SHA-256', new TextEncoder().encode(password));
    return encode(new Uint8Array(hash));
  }
  return hashPasswordJS(password);
}

export function saveBlob(blob: EncryptedBlob): void {
  localStorage.setItem(SECRETS_STORAGE_KEY, JSON.stringify(blob));
}

export function loadBlob(): EncryptedBlob | null {
  try {
    const raw = localStorage.getItem(SECRETS_STORAGE_KEY);
    if (!raw) return null;
    const parsed = JSON.parse(raw);
    if (parsed.iv && parsed.data && parsed.salt && parsed.hash) {
      return parsed;
    }
    return null;
  } catch {
    return null;
  }
}

export function clearBlob(): void {
  localStorage.removeItem(SECRETS_STORAGE_KEY);
}

File: script-dashboard/python-server/crypto/secrets.py
"""Secrets encryption using PBKDF2 + AES-GCM 256-bit."""
import os
import json
import hashlib
from base64 import b64encode, b64decode

from Crypto.Protocol.KDF import PBKDF2
from Crypto.Cipher import AES
from Crypto.Random import get_random_bytes
from Crypto.Hash import SHA256

PBKDF2_ITERATIONS = 200000
PBKDF2_KEY_LEN = 32  # 256 bits
AES_KEY_LEN = 32
IV_LEN = 12
SALT_LEN = 16


def _derive_key(password: str, salt: bytes) -> bytes:
    """Derive AES key from password using PBKDF2."""
    return PBKDF2(
        password.encode("utf-8"),
        salt,
        dkLen=PBKDF2_KEY_LEN,
        count=PBKDF2_ITERATIONS,
        hmac_hash_module=SHA256,
    )


def encrypt_secrets(values: dict, password: str) -> dict:
    """Encrypt a secrets dict with password. Returns blob dict."""
    if not values:
        values = {}

    salt = get_random_bytes(SALT_LEN)
    key = _derive_key(password, salt)
    iv = get_random_bytes(IV_LEN)

    plaintext = json.dumps(values).encode("utf-8")

    cipher = AES.new(key, AES.MODE_GCM, nonce=iv)
    ciphertext, tag = cipher.encrypt_and_digest(plaintext)

    # Combine ciphertext + tag for transport
    data = b64encode(ciphertext + tag).decode("ascii")

    # Store password hash for verification (SHA-256 of password)
    pw_hash = hashlib.sha256(password.encode("utf-8")).hexdigest()

    return {
        "iv": b64encode(iv).decode("ascii"),
        "data": data,
        "salt": b64encode(salt).decode("ascii"),
        "hash": pw_hash,
    }


def decrypt_secrets(blob: dict, password: str) -> dict:
    """Decrypt a secrets blob with password. Returns values dict."""
    salt = b64decode(blob["salt"])
    iv = b64decode(blob["iv"])
    data = b64decode(blob["data"])

    key = _derive_key(password, salt)

    # Split ciphertext and tag (tag is last 16 bytes)
    tag_len = 16
    ciphertext = data[:-tag_len]
    tag = data[-tag_len:]

    cipher = AES.new(key, AES.MODE_GCM, nonce=iv)
    plaintext = cipher.decrypt_and_verify(ciphertext, tag)

    return json.loads(plaintext.decode("utf-8"))


def hash_password(password: str) -> str:
    """Return SHA-256 hex digest of password for verification."""
    return hashlib.sha256(password.encode("utf-8")).hexdigest()
